### ISSUE TITLE
Cloning bitcoin-s repo takes forever

### ISSUE THREAD

**Issue by benthecarman** [bitcoin-s/bitcoin-s#2830]
Doing a `git clone` of the repo takes a very long time. Is there something we can do to reduce the size of the repo besides just squashing commits?

**Comment by Christewart** [comment]
Related to #2674 

So i did some googling on this this morning, and found this stackoverflow post

https://stackoverflow.com/questions/10622179/how-to-find-identify-large-commits-in-git-history

I ran the script the provided and added the filter to check elements that are not in `HEAD` (`grep -vF --file=<(git ls-tree -r HEAD | awk '{print $3}') |`)


```
git rev-list --objects --all |
  git cat-file --batch-check='%(objecttype) %(objectname) %(objectsize) %(rest)' |
  sed -n 's/^blob //p' |
  grep -vF --file=<(git ls-tree -r HEAD | awk '{print $3}') |
  sort --numeric-sort --key=2 |
  cut -c 1-12,41- |
  $(command -v gnumfmt || echo numfmt) --field=2 --to=iec-i --suffix=B --padding=7 --round=nearest
```

and I get a huge list of seemingly all of our html files 

>0da968cbbb0a  1.1MiB api/org/bitcoins/testkit/chain/ChainDbUnitTest.html
117b43ca63e2  1.1MiB api/org/bitcoins/testkit/chain/ChainDbUnitTest.html
251e55c15503  1.1MiB api/org/bitcoins/testkit/chain/ChainDbUnitTest.html
273cd58d2d00  1.1MiB api/org/bitcoins/testkit/chain/ChainDbUnitTest.html
27b34daf1924  1.1MiB api/org/bitcoins/testkit/chain/ChainDbUnitTest.html
2fc75e867421  1.1MiB api/org/bitcoins/testkit/chain/ChainDbUnitTest.html
...

So i'm wondering if we are redundantly committing all these files on every merge? Idk

**Comment by benthecarman** [comment]
These are generated from unidoc, I believe we should ignore those

**Comment by Christewart** [comment]
I'm suspicious of this code, as it seems to be moving the unidoc stuff to another place, which means it isn't in the `.gitignore` any more

https://github.com/bitcoin-s/bitcoin-s/blob/478ab16e28c65a4796f92433a6fef12cd35e351a/build.sbt#L281

**Comment by Christewart** [comment]
So i've been playing around with the [`filter-repo`](https://github.com/newren/git-filter-repo) tool for git. Some handy documentation for how this tool can be used 

https://htmlpreview.github.io/?https://github.com/newren/git-filter-repo/blob/docs/html/git-filter-repo.html#EXAMPLES

It has a really handy flag called `--analyze` which gives you state of the world with the git repo. 

I've attached the files that are output when running 

```
$ git filter-repo --analyze   
```

which is located are generated to `.git/filter-repo/analysis`. I've omitted one file (`blob-shas-and-paths.txt`) as it is too big to attach to github

[path-all-sizes.txt](https://github.com/bitcoin-s/bitcoin-s/files/6246202/path-all-sizes.txt)
[path-deleted-sizes.txt](https://github.com/bitcoin-s/bitcoin-s/files/6246203/path-deleted-sizes.txt)
[extensions-all-sizes.txt](https://github.com/bitcoin-s/bitcoin-s/files/6246204/extensions-all-sizes.txt)
[extensions-deleted-sizes.txt](https://github.com/bitcoin-s/bitcoin-s/files/6246205/extensions-deleted-sizes.txt)
[directories-all-sizes.txt](https://github.com/bitcoin-s/bitcoin-s/files/6246206/directories-all-sizes.txt)
[directories-deleted-sizes.txt](https://github.com/bitcoin-s/bitcoin-s/files/6246207/directories-deleted-sizes.txt)
[renames.txt](https://github.com/bitcoin-s/bitcoin-s/files/6246208/renames.txt)

If you open up `path-deleted-sizes.txt` you will see things like that look like this (note the `api/` prefix)

```
=== Deleted paths by reverse accumulated size ===
Format: unpacked size, packed size, date deleted, path name(s)
   200655131    7159648 2021-03-02 api/org/bitcoins/testkit/core/gen/ScriptGenerators$.html
   188719101    6566535 2021-03-02 api/org/bitcoins/testkit/core/gen/TransactionGenerators$.html
   180024860    5580489 2021-03-02 api/org/bitcoins/testkit/core/gen/ScriptGenerators.html
    66837986    3516479 2021-03-02 api/org/bitcoins/testkit/util/TransactionTestUtil$.html
    90428536    3363848 2021-03-02 api/org/bitcoins/testkit/core/gen/NumberGenerator$.html
    79142752    3236881 2021-03-02 api/org/bitcoins/testkit/core/gen/NumberGenerator.html
    61384967    3160720 2021-03-02 api/org/bitcoins/testkit/util/TransactionTestUtil.html
    16627908    2828933 2021-03-24 secp256k1jni/natives/linux_64/libsecp256k1.a
...
```

**Comment by Christewart** [comment]
So i've got another repo setup to that filters out all of the `api/` files that were identified by `git filter-repo --analyze`

The command: 

> git filter-repo --path api/ --invert-paths 

https://github.com/Christewart/bitcoin-s-git-rework

It does re-write the entire git history, for instance these two commits are equivalent

1. https://github.com/bitcoin-s/bitcoin-s/commit/2745fdf8db9b1ae254a5f844a1c599a427bafb3a
2. https://github.com/Christewart/bitcoin-s-git-rework/commit/8588ce77ed82059a0f17573ff9b2caeb387b3c5d

So this invites the question of when to do this, to reduce repo size. It would probably be best to do this after 

1. We pull down adaptor-dlc work into master
2. We cut a release 

Now that we have core scalajs components in, it may be worth releasing, but it seems we would really want to wait until adaptor-dlc is merged too.

Thoughts?

**Comment by nkohen** [comment]
We should hash the directories (at least for source code) and compare those hashes to current `master` to check for changes. If there aren't any then I think that there is no harm in merging whenever as cherry-picking will not run into any conflicts

**Comment by Christewart** [comment]
So responding to this comment on another issue: https://github.com/bitcoin-s/bitcoin-s/pull/2880#discussion_r609856573

I did try to run the website on https://github.com/Christewart/bitcoin-s-git-rework/commit/8588ce77ed82059a0f17573ff9b2caeb387b3c5d and access the website's scaladocs and things seemed to work

**Comment by Christewart** [comment]
So today i force pushed to clean up the git history as detailed above. However it seems  to not work with the default 

>git clone git@github.com:bitcoin-s/bitcoin-s.git

however it DOES work if we just do the master branch: 

>git clone -b master --single-branch git@github.com:bitcoin-s/bitcoin-s.git


I think we need rebase these branches on origin to get the full effect

https://github.com/bitcoin-s/bitcoin-s/tree/adaptor-dlc

https://github.com/bitcoin-s/bitcoin-s/tree/hodl-mccormack-dlc

**Comment by Christewart** [comment]
Ok, another problem was the `gh-pages` branch. I filtered out old content on that branch as well. 

This issue is being closed on e7d34a9ba93c07a762ab7eae7b9d0a83422cc526

and the time to do a fresh clone is 

```
time git clone git@github.com:bitcoin-s/bitcoin-s.git                                                                    ✔  14454  12:43:48
Cloning into 'bitcoin-s'...
remote: Enumerating objects: 188916, done.
remote: Counting objects: 100% (30920/30920), done.
remote: Compressing objects: 100% (6470/6470), done.
remote: Total 188916 (delta 13444), reused 30844 (delta 13434), pack-reused 157996
Receiving objects: 100% (188916/188916), 54.15 MiB | 2.04 MiB/s, done.
Resolving deltas: 100% (86793/86793), done.
git clone git@github.com:bitcoin-s/bitcoin-s.git  11.82s user 1.26s system 41% cpu 31.884 total

```


>11.82s user 1.26s system 41% cpu 31.884 total
